clients/src/test/java/org/apache/kafka/common/requests/OffsetFetchResponseTest.java

```diff
@@ -0,0 +1,169 @@
+            leaderEpochOne,
+            metadata,
+            Errors.TOPIC_AUTHORIZATION_FAILED
+        ));
+        partitionDataMap.put(new TopicPartition(topicTwo, partitionTwo), new PartitionData(
+            offset,
+            leaderEpochTwo,
+            metadata,
+            Errors.UNKNOWN_TOPIC_OR_PARTITION
+        ));
+    }
+
+    @Test
+    public void testConstructor() {
+        OffsetFetchResponse response = new OffsetFetchResponse(throttleTimeMs, Errors.NOT_COORDINATOR, partitionDataMap);
+        assertEquals(Errors.NOT_COORDINATOR, response.error());
+        assertEquals(Collections.singletonMap(Errors.NOT_COORDINATOR, 1), response.errorCounts());
+
+        assertEquals(throttleTimeMs, response.throttleTimeMs());
+
+        Map<TopicPartition, PartitionData> responseData = response.responseData();
+        assertEquals(partitionDataMap, responseData);
+        responseData.forEach(
+            (tp, data) -> assertTrue(data.hasError())
+        );
+    }
+
+    @Test
+    public void testCouldPopulatePartitionError() {
+        partitionDataMap.put(new TopicPartition(topicTwo, partitionTwo), new PartitionData(
+            offset,
+            leaderEpochTwo,
+            metadata,
+            Errors.GROUP_AUTHORIZATION_FAILED
+        ));
+
+        OffsetFetchResponse latestResponse = new OffsetFetchResponse(throttleTimeMs, Errors.NONE, partitionDataMap);
+        Struct versionOneStruct = latestResponse.data.toStruct((short) 1);
+        assertFalse(versionOneStruct.hasField(ERROR_CODE));
+
+        OffsetFetchResponse oldResponse =  new OffsetFetchResponse(versionOneStruct, (short) 1);
+
+        // Partition level error populated in older versions.
+        assertEquals(Errors.GROUP_AUTHORIZATION_FAILED, oldResponse.error());
+        assertEquals(Collections.singletonMap(Errors.GROUP_AUTHORIZATION_FAILED, 1), oldResponse.errorCounts());
+
+        assertEquals(DEFAULT_THROTTLE_TIME, oldResponse.throttleTimeMs());
+
+        Map<TopicPartition, PartitionData> responseData = oldResponse.responseData();
+        Map<TopicPartition, PartitionData> expectedDataMap = new HashMap<>();
+        for (Map.Entry<TopicPartition, PartitionData> entry : partitionDataMap.entrySet()) {
+            PartitionData partitionData = entry.getValue();
+            expectedDataMap.put(entry.getKey(), new PartitionData(
+                partitionData.offset,
+                Optional.empty(), // Leader epoch is not set in version < 5
+                partitionData.metadata,
+                partitionData.error
+            ));
+        }
+        assertEquals(expectedDataMap, responseData);
+
+        responseData.forEach(
+            (tp, data) -> assertTrue(data.hasError())
+        );
+    }
+
+    @Test
+    public void testShouldThrottle() {
+        OffsetFetchResponse response = new OffsetFetchResponse(throttleTimeMs, Errors.NONE, partitionDataMap);
+        assertFalse(response.shouldClientThrottle((short) 3));
```

Comment:
It would be better to write these test cases without relying on a specific latest version. We can use `ApiKeys.latestVersion` instead so that the test case automatically covers new versions.